standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java

```diff
@@ -5394,6 +5410,476 @@ public void countOpenTxns() throws MetaException {
     }
   }
 
+  private void cleanOldStatsFromPartColStatTable(Map<PartitionInfo, ColumnStatistics> statsPartInfoMap,
+                                                 Connection dbConn) throws SQLException {
+    PreparedStatement preparedStatement = null;
+    int numRows = 0;
+    int maxNumRows = MetastoreConf.getIntVar(conf, ConfVars.DIRECT_SQL_MAX_ELEMENTS_VALUES_CLAUSE);
+
+    // Index is present on DB_NAME,TABLE_NAME,COLUMN_NAME,PARTITION_NAME. use that.
+    // TODO : Need to add catalog name to the index
+    String delete = "DELETE FROM \"PART_COL_STATS\" where \"DB_NAME\" = ? AND "
+            + "\"TABLE_NAME\" = ? AND \"COLUMN_NAME\" = ? AND \"PARTITION_NAME\" = ? "
+            + "AND \"PART_ID\" = ?";
+
+    try {
+      preparedStatement = sqlGenerator.prepareStmtWithParameters(dbConn, delete, null);
+      for (Map.Entry entry : statsPartInfoMap.entrySet()) {
+        ColumnStatistics colStats = (ColumnStatistics) entry.getValue();
+        PartitionInfo partitionInfo = (PartitionInfo) entry.getKey();
+        for (ColumnStatisticsObj statisticsObj : colStats.getStatsObj()) {
+          preparedStatement.setString(1, colStats.getStatsDesc().getDbName());
+          preparedStatement.setString(2, colStats.getStatsDesc().getTableName());
+          preparedStatement.setString(3, statisticsObj.getColName());
+          preparedStatement.setString(4, colStats.getStatsDesc().getPartName());
+          preparedStatement.setLong(5, partitionInfo.partitionId);
+          numRows++;
+          preparedStatement.addBatch();
+          if (numRows == maxNumRows) {
+            preparedStatement.executeBatch();
+            numRows = 0;
+            LOG.debug("Executed delete " + delete + " for numRows " + numRows);
+          }
+        }
+      }
+
+      if (numRows != 0) {
```

Comment:
> 0 ?

Reply:
in this case it does not matter i think.

Reply:
it's more for readablility